pkg/cmd/roachtest/decommission.go

```diff
@@ -257,34 +273,92 @@ func execCLI(
 	return string(buf), err
 }
 
-func runDecommissionAcceptance(ctx context.Context, t *test, c *cluster) {
+// runDecommissionRecommission tests a bunch of node
+// decommissioning/recommissioning procedures, all the while checking for
+// replica movement and appropriate membership status detection behavior. We go
+// through partial decommissioning of random nodes, ensuring we're able to undo
+// those operations. We then fully decommission nodes, verifying it's an
+// irreversible operation.
+func runDecommissionRecommission(ctx context.Context, t *test, c *cluster) {
 	args := startArgs("--env=COCKROACH_SCAN_MAX_IDLE_TIME=5ms")
 	c.Put(ctx, cockroach, "./cockroach")
 	c.Start(ctx, t, args)
 
+	// We use a few helpers for random node ID access.
+	var nodeIDs []int
+	for i := 1; i <= c.spec.NodeCount; i++ {
+		nodeIDs = append(nodeIDs, i)
+	}
+	getRandNode := func() int {
+		return nodeIDs[rand.Intn(len(nodeIDs))]
+	}
+	getRandNodeOtherThan := func(ids ...int) int {
+		for {
+			cur := nodeIDs[rand.Intn(len(nodeIDs))]
+			inBlockList := false
+			for _, id := range ids {
+				if cur == id {
+					inBlockList = true
+				}
+			}
+			if inBlockList {
+				continue
+			}
+			return cur
+		}
+	}
+
+	// We define a few short hands to {d,r}ecommission target nodes, through a
+	// given node.
 	decommission := func(
 		ctx context.Context,
+		targetNodes nodeListOption,
 		runNode int,
+		verbs ...string,
+	) (string, error) {
+		args := []string{"node", "decommission"}
+		args = append(args, verbs...)
+
+		if len(targetNodes) == 1 && targetNodes[0] == runNode {
+			args = append(args, "--self")
+		} else {
+			for _, target := range targetNodes {
+				args = append(args, strconv.Itoa(target))
+			}
+		}
+		return execCLI(ctx, t, c, runNode, args...)
+	}
+	recommission := func(
+		ctx context.Context,
 		targetNodes nodeListOption,
+		runNode int,
 		verbs ...string,
 	) (string, error) {
-		args := []string{"node"}
+		args := []string{"node", "recommission"}
 		args = append(args, verbs...)
-		for _, target := range targetNodes {
-			args = append(args, strconv.Itoa(target))
+
+		if len(targetNodes) == 1 && targetNodes[0] == runNode {
+			args = append(args, "--self")
+		} else {
+			for _, target := range targetNodes {
+				args = append(args, strconv.Itoa(target))
+			}
 		}
 		return execCLI(ctx, t, c, runNode, args...)
 	}
 
-	getCsvNumCols := func(csvStr string) (cols int, err error) {
+	// getCsvNumCols returns the number of columns in the given csv string.
+	getCsvNumCols := func(csvStr string) (cols int) {
 		reader := csv.NewReader(strings.NewReader(csvStr))
 		records, err := reader.Read()
 		if err != nil {
-			return 0, errors.Errorf("error reading csv input: \n %v\n errors:%s", csvStr, err)
+			t.Fatal(errors.Errorf("error reading csv input: \n %v\n errors:%s", csvStr, err))
 		}
-		return len(records), nil
+		return len(records)
 	}
 
+	// matchCsv matches a multi-line csv string with the provided regex
```

Comment:
That we do all of this matching here is sort of crazy and makes this test a real mess. Ideally we'd have the decommission command a super thin wrapper around hitting a gRPC endpoint. Definitely not something you can just drive-by here though.

Curious if @knz has an intermediate idea. The hoops we jump through in this test are crazy, no? This test would be so much less messy if we just cared about the exit status of the invocation rather than its output.